src/shims/windows/handle.rs

```diff
@@ -0,0 +1,162 @@
+use rustc_target::abi::HasDataLayout;
+use std::mem::variant_count;
+
+use crate::*;
+
+/// A Windows `HANDLE` that represents a resource instead of being null or a pseudohandle.
+///
+/// This is a seperate type from [`Handle`] to simplify the packing and unpacking code.
+#[derive(Clone, Copy)]
+enum RealHandle {
+    Thread(ThreadId),
+}
+
+impl RealHandle {
+    const USABLE_BITS: u32 = 31;
+
+    const THREAD_DISCRIMINANT: u32 = 1;
+
+    fn discriminant(self) -> u32 {
+        match self {
+            // can't use zero here because all zero handle is invalid
+            Self::Thread(_) => Self::THREAD_DISCRIMINANT,
+        }
+    }
+
+    fn data(self) -> u32 {
+        match self {
+            Self::Thread(thread) => thread.to_u32(),
+        }
+    }
+
+    fn packed_disc_size() -> u32 {
+        // log2(x) + 1 is how many bits it takes to store x
```

Comment:
For x=4, we only need 2 bits, but this will say 3... so it doesn't seem quite optimal?